language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala

```diff
@@ -0,0 +1,333 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.codegen.backend.java.inner
+
+import com.daml.ledger.javaapi
+import com.daml.lf.codegen.backend.java.{JavaEscaper, ObjectMethods}
+import com.daml.lf.data.ImmArray.ImmArraySeq
+import com.daml.lf.data.Ref.{ChoiceName, PackageId, QualifiedName}
+import com.daml.lf.iface.{
+  DefInterface,
+  PrimType,
+  TemplateChoice,
+  Type,
+  TypeCon,
+  TypeNumeric,
+  TypePrim,
+  TypeVar,
+}
+import com.squareup.javapoet._
+import com.typesafe.scalalogging.StrictLogging
+
+import java.util.Optional
+import javax.lang.model.element.Modifier
+import scala.jdk.CollectionConverters._
+
+object InterfaceClass extends StrictLogging {
+
+  def classNameForInterface(qualifiedName: QualifiedName) =
+    ClassName.bestGuess {
+      val QualifiedName(module, name) = qualifiedName
+      // consider all but the last name segment to be part of the java package name
+      val packageSegments = module.segments.slowAppend(name.segments).toSeq.dropRight(1)
+      // consider the last name segment to be the java class name
+      val className = name.segments.toSeq.takeRight(1)
+
+      val packageName = packageSegments.map(_.toLowerCase)
+
+      (packageName ++ className)
+        .filter(_.nonEmpty)
+        .map(JavaEscaper.escapeString)
+        .mkString(".")
+    }
+
+  def generate(
+      interfaceName: ClassName,
+      interface: DefInterface.FWT,
+      packagePrefixes: Map[PackageId, String],
+      packageId: PackageId,
+      interfaceId: QualifiedName,
+  ): TypeSpec =
+    TrackLineage.of("interface", interfaceName.simpleName()) {
+      logger.info("Start")
+
+      val templateType = TypeSpec
+        .classBuilder(interfaceName)
+        .addModifiers(Modifier.FINAL, Modifier.PUBLIC)
+        .addField(generateTemplateIdField(packageId, interfaceId))
+        .addType(
+          generateIdClass(
+            interfaceName,
+            interface.choices,
+            packagePrefixes,
+          )
+        )
+        .addType(generateContractClass(interfaceName))
+        .build()
+      logger.debug("End")
+      templateType
+    }
+
+  private val idFieldName = "id"
```

Comment:
Part of this stuff here badly needs some refactoring out into something else because this is mostly duplicating code from `TemplateClass.scala`. I need to figure out first though, which parts are easy to share between both classes, because the duplicated methods in here are slightly but importantly different.